### PR TITLE
feat: usize for array types

### LINKED ISSUE #4654: RFC: Array.usize
### Proposal

I propose to add `Array.usize` to get the `USize` size of an array, without going through `Nat`.

Currently, to get the `USize` of an array, one needs to use `Array.size` and then `Nat.toUSize`. The latter takes the return value of `Array.size` and calculates the remainder modulo `USize.size`. This is not optimal since arrays are stored in memory and therefore cannot exceed `USize.size`.  This proposal would bypass the remainder step and return the correct array size as an `USize` scalar (which is trivial to access in the current array implementation).

Other array operations have optimizations that bypass this trivial size check for the sake of nearly matching C-language performance. Some of my code at `UnicodeBasic` and elsewhere could benefit a lot from this nearly trivial improvement to the array API.

The drawback is that this adds one more unsafe implementation. So this request needs some careful consideration.

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by fgdorais**
Add efficient `usize` functions for `Array`, `ByteArray`, `FloatArray`.

This is part 1 of 2 since there is a need to update stage0 between the two parts. (See discussion below.)

Closes #4654

### CODE REVIEW

**Comment by fgdorais** [comment]
It was necessary to update stage0 in order to use the `usize` functions.

**Comment by fgdorais** [comment]
awaiting-review

**Comment by nomeata** [comment]
I didn't expect this to need a stage0 update. Why is that so?

**Comment by fgdorais** [comment]
I didn't expect that either. It's because these are `extern` definitions so there is no code for them until stage0 is updated.

**Comment by nomeata** [comment]
Hmm, still somewhat strange, the the C symbol should already be there, shouldn’t it? Did you try setting `prefer_native` to true?

But I’m no expert here. I’m just bringing this up because PRs with stage0 updates need to be manually merged by (I think) Sebastian or Leo. If you’d split it in two PRs then more people can merge and it can go through the normal merge queue.

**Comment by fgdorais** [comment]
I'm not an expert there either. If I don't update stage0 then calls to `usize` look for the reference implementation instead of the extern symbol, but the reference implementation is never compiled.

Yes, I read the manual. I also made sure the commits were clean. I was thinking about splitting in two but I was hoping Sebastian would say something before I did that. Either way, there needs to be a stage0 update between the two commits.

**Comment by nomeata** [comment]
Right, this works of course! It’s just that with two PRs, the stage0 update can happen via the github bot. But not problem for me letting Sebastian merge this PR :-)

**Comment by fgdorais** [comment]
There's a bot for that! I thought Sebastian would have to do it either way...

I will split right away then!

**Comment by nomeata** [comment]
There is a bot these days. It runs automatically after a PR merge whenever it touches `stage0/src/stdlib_flags.h`, or anyone with commit rights can trigger it.

So you could even leave a comment in `stage0/src/stdlib_flags.h` of the form
```
// please update stage0
```
and it’ll happen automatically after the PR enters master (and it will nicely override this comment).

File: src/Init/Data/Array/Basic.lean
Comment (nomeata) [inline src/Init/Data/Array/Basic.lean:58]: It seems that other functions from C world that are used via `extern` have a
`LEAN_EXPORT` declaration in the corresponding header file. Did you intentionally not add one of these for the new functions?

Comment (fgdorais) [reply src/Init/Data/Array/Basic.lean:58]: I think `LEAN_EXPORT` only matters for symbols exported to a shared library. That doesn't seem necessary in this case. For example, `lean_array_get_size` (which implements `Array.size`) does not have a `LEAN_EXPORT`.

Comment (nomeata) [reply src/Init/Data/Array/Basic.lean:58]: Great, thanks for checking.

**Review by nomeata — APPROVED** [review]

**Comment by fgdorais** [comment]
Thank you so much for your help Joachim! This PR would have stayed in limbo for a long time without your enlightening comments!

**Pull request merged**
